e3nn/util/codegen/_einsum.py

```diff
@@ -0,0 +1,141 @@
+"""Optimize einsums based on shapes for code generation."""
+
+from typing import List, Optional
+
+from opt_einsum.contract import contract_path
+
+from e3nn.util import prod
+
+
+# based on opt_einsum/backends/torch.py
+class _CodeGenBackend:
+    def __init__(self):
+        self._tmp_idx = 0
+        self.code = []
+
+    def get_tmp(self):
+        out = f"_ein_tmp{self._tmp_idx}"
+        self._tmp_idx += 1
+        return out
+
+    def transpose(self, v, axes, out=None):
+        axes = list(axes)
+        if axes == list(range(len(axes))):
+            # In this case, we're asking for the identity permutation, so just no-op
+            if out is None:
+                return v
+            else:
+                self.code.append(f"{out} = {v}  # identity permutation no-op")
+        out = out if out is not None else self.get_tmp()
+        self.code.append(f"{out} = {v}.permute({', '.join(str(a) for a in axes)})")
+        return out
+
+    def einsum(self, einstr, *args, out=None):
+        out = out if out is not None else self.get_tmp()
```

Comment:
If I understand you right, the difference between this and the `transpose` example is that for `transpose` its an identity operation, but for einsum we're always doing something and need somewhere to put it.

Since we're only passing around names of objects inside the codegen, everything needs a name — nothing can be implicit. And even though I'm sure I could reuse a bunch of those temporary variables, I thought it was less error prone to just always make new ones and let the TorchScript compiler hopefully notice that they are unused later and delete them

We could always add explicit `del _ein_tmp_whatever` statements in the codegen?

Reply:
Forget about it, I simply did not understand it. 

I thought one coud do something like

v = "var"
v = codegen.mul(v, 4.0)

then v is now "var.mul(4.0)"

But we don't want that